### PR TITLE
ENH: add strip methods to moltype

### LINKED ISSUE #1921: Sequence and SeqView classes should not be used to replace characters
Delete the `_seq_filter()` method from sequence classes in `new_sequence.py`. No longer being used.

Also delete `new_sequence.Sequence.replace()` method.  Users will need to do this explicitly themselves.

There will be side-effects in that some functions (e.g. `moltype._convert_to_rna()` and `_convert_to_dna()`) employ the replace method. We can just produce revised functions that cast incoming data to string and using the replace method on that.

### PR DESCRIPTION
**Pull request by KatherineCaley**
- add `strip_degenerate`, `strip_bad`, `strip_bad_and_gaps`, `disambiguate` to `Moltype` -- addresses #1932
- add `count_variants`, `count_gaps`, `count_degenerate` to `Moltype` -- addresses #1938
- remove replace from sequences -- fixes #1921 
- port Sequence tests

### CODE REVIEW

File: src/cogent3/core/new_sequence.py [hunk 1]
```diff
@@ -327,84 +329,80 @@ def strip_degenerate(self):
         """Removes degenerate bases by stripping them out of the sequence."""
         return self.__class__(
             moltype=self.moltype,
-            seq=self.moltype.strip_degenerate(self),
+            seq=self.moltype.strip_degenerate(str(self)),
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:332]: comment about refactor, perhaps `numpy.array(self)` more efficient?

File: src/cogent3/core/new_sequence.py [hunk 2]
```diff
@@ -327,84 +329,80 @@ def strip_degenerate(self):
...
             info=self.info,
         )
 
     def strip_bad(self):
         """Removes any symbols not in the alphabet."""
+        # refactor: design
+        # previous implementation did NOT strip X from DNA even though its not in the alphabet
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:339]: no, we don't want to keep that behaviour

File: src/cogent3/core/new_sequence.py [hunk 3]
```diff
@@ -327,84 +329,80 @@ def strip_degenerate(self):
...
-            moltype=self.moltype, seq=self.moltype.strip_bad(self), info=self.info
+            moltype=self.moltype, seq=self.moltype.strip_bad(str(self)), info=self.info
         )
 
     def strip_bad_and_gaps(self):
         """Removes any symbols not in the alphabet, and any gaps."""
         return self.__class__(
             moltype=self.moltype,
-            seq=self.moltype.strip_bad_and_gaps(self),
+            seq=self.moltype.strip_bad_and_gaps(str(self)),
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:349]: add comment on same issue as above: `str(self)` versus `numpy.array(self)`

File: src/cogent3/core/new_sequence.py [hunk 4]
```diff
@@ -327,84 +329,80 @@ def strip_degenerate(self):
...
         )
 
-    def is_gapped(self):
+    def is_gapped(self) -> bool:
         """Returns True if sequence contains gaps."""
-        return self.moltype.is_gapped(self)
+        return self.moltype.is_gapped(str(self))
 
-    def is_gap(self, char=None):
+    def is_gap(self, char: str = None) -> bool:
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:357]: checking how often this method used, I suspect we can just delete this method and associated tests

File: src/cogent3/core/new_sequence.py [hunk 5]
```diff
@@ -1113,34 +1117,51 @@ def add_feature(
...
+        else:
+            # assume converting between ASCII/BYTES and nucleic acid
+            seq = str(self)
+
+        if not moltype.most_degen_alphabet().is_valid(seq):
+            raise ValueError(
+                f"Changing from old moltype={self.moltype.label} to new "
+                f"moltype={moltype.label} is not valid for this data"
+            )
+        sv = SeqView(seq=seq, alphabet=moltype.most_degen_alphabet())
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:1160]: add a comment to the docstring that if this method is applied to a sequence that is part of a sequence collection, the result is no longer part of that collection.

File: src/cogent3/core/new_sequence.py [hunk 6]
```diff
@@ -1699,7 +1711,7 @@ def complement(self):
         will return list of keys.
         """
         return self.__class__(
-            moltype=self.moltype, seq=self.moltype.complement(self), info=self.info
+            moltype=self.moltype, seq=self.moltype.complement(str(self)), info=self.info
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:1714]: everywhere you are using `str(self)` add a comment about checking using either `array(self)`.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
massive amount of work!! mainly trivial comments to be added. One thing to note is that everywhere you use `str(self)` on some objects, `numpy.array` is probably available too. Se we should add a comment about checking performance of those.

File: src/cogent3/core/new_sequence.py (hunk 1 above)
Comment (KatherineCaley) [reply src/cogent3/core/new_sequence.py:332]: the underlying methods uses `bytes` so i've changed it to that

**Pull request merged**